### PR TITLE
LanguageService - getSuggestionDiagnostics, getEditsForRefactor, getCodeFixesAtPosition

### PR DESCRIPTION
**Pull request by cancerberoSgx**
The first commit implements LanguageService getEditsForRefactor and getCodeFixesAtPosition and wraps all the data types used by those. 

The last commit shows how this can be useful to (re)use exsting TypeScript refactors in ts-simple-ast. In this case I implemented ArrowFunction.addBracesToBody and ArrowFunction.removeBracesToBody just using existing typescript code fix `"Add or remove braces in an arrow function"`. And also implemented Statement.moveToNewFile, using existing typescript refactor `"Move to a new file"`. 

I think TypeScript team slowly starts to focus on providing useful refactoring tools and I think ts-simple-ast should use these to enrich its high level APIs. I can keep adding more refactors provided by TS but first I wanted to ask you what do you think.

### CODE REVIEW

File: src/compiler/ast/function/ArrowFunction.ts [hunk 1]
```diff
@@ -13,4 +12,30 @@ export class ArrowFunction extends ArrowFunctionBase<ts.ArrowFunction> {
     getEqualsGreaterThan(): Node<ts.Token<SyntaxKind.EqualsGreaterThanToken>> {
         return this.getNodeFromCompilerNode(this.compilerNode.equalsGreaterThanToken);
     }
+
+    /**
+     * Add braces to this arrow function body in case is viable.
+     *
+     * WARNING! This will forget all the nodes in the file! It's best to do this after you're all done with the file.
```
Comment (dsherret) [inline src/compiler/ast/function/ArrowFunction.ts]: Could you remove this from the PR? There are some refactors implemented in `getEditsForRefactor` that we can do a better job at that won't cause forgetting all the nodes.

File: src/compiler/ast/statement/Statement.ts [hunk 2]
```diff
@@ -9,4 +10,27 @@ export class Statement<T extends ts.Statement = ts.Statement> extends Node<T> {
...
+     * Will move the statement to a new file updating all the references in the project.
+     *
+     * If `nodeFilePath` is not provided then the name of the new file will be the one of the first symbol
+     * found or `newFile` if none is found.
+     *
+     * WARNING! This will forget all the nodes in this node's file! It's best to do this after you're all done with the file.
+     *
+     * @returns the new SourceFile to which the statement was moved or undefined if the statement coulnd't be moved.
+     */
+    moveToNewFile(newFilePath?: string): SourceFile | undefined {
```
Comment (dsherret) [inline src/compiler/ast/statement/Statement.ts]: Could you remove this method from the PR? I don't want to suggest these refactors by providing an easy way to do them on the objects because they cause nodes to be forgotten.

Comment (dsherret) [reply src/compiler/ast/statement/Statement.ts]: Perhaps we could add them some other way in the future, but I don't think it should be so readily available directly on `Statement`

File: src/compiler/tools/LanguageService.ts [hunk 3]
```diff
@@ -337,6 +340,41 @@ export class LanguageService {
...
+    /**
+     * Gets the edits information to modify the file with given refactor action at given position.
+     * @param fileName - File path of the source file.
+     * @param formatOptions - Fomat code settings.
+     * @param positionOrRange - The position in the source file where to apply given refactor.
+     * @param refactorName  - The refactor name.
+     * @param actionName  - The refactor action name.
+     * @param preferences - User preferences for refactoring.
+     */
+    getEditsForRefactor(fileName: string, formatOptions: FormatCodeSettings, positionOrRange: number | TextRange,
```
Comment (dsherret) [inline src/compiler/tools/LanguageService.ts]: What do you think about making this a single object parameter with all of these parameters as properties on the object?

Comment (dsherret) [reply src/compiler/tools/LanguageService.ts]: (and same for `getCodeFixesAtPosition`)

File: src/compiler/tools/results/TextChange.ts [hunk 4]
```diff
@@ -9,7 +9,6 @@ export class TextChange {
     /** @internal */
     private readonly _compilerObject: ts.TextChange;
 
-    /** @internal */
```
Comment (dsherret) [inline src/compiler/tools/results/TextChange.ts]: Could you make this internal again? Also, add `/** @internal **/` to the constructors of all the other classes.

In the future, I'm thinking to change these to have private modifiers in the declaration file.

File: src/tests/compiler/tools/languageServiceTests.ts [hunk 5]
```diff
@@ -112,6 +112,57 @@ describe(nameof(LanguageService), () => {
...
+
+    describe(nameof<LanguageService>(l => l.getEditsForRefactor), () => {
+        it("should get edits for known refactor 'move to a new file'", () => {
+            const { sourceFile, project } = getInfoFromText("export class A {}; function f(){return new A(); }", { filePath: "/file.ts" });
+            const classId = sourceFile.getClassOrThrow("A").getNameNodeOrThrow();
+            const results = project.getLanguageService().getEditsForRefactor(sourceFile.getFilePath(), {},
+                { pos: classId.getStart(), end: classId.getEnd() }, "Move to a new file", "Move to a new file", {});
+
+            const edit1 = results!.getEdits().find(edit => edit.getFilePath() === sourceFile.getFilePath());
+            const edit2 = results!.getEdits().find(edit => edit.getFilePath() === "/A.ts");
```
Comment (dsherret) [inline src/tests/compiler/tools/languageServiceTests.ts]: Does edits return more than two results? Could you add an assertion for how many results it should return?

File: src/tests/compiler/tools/languageServiceTests.ts [hunk 6]
```diff
@@ -112,6 +112,57 @@ describe(nameof(LanguageService), () => {
...
+    describe(nameof<LanguageService>(l => l.getCodeFixesAtPosition), () => {
+        it("should get code fixes at position for known code fixes convertToEs6Module (error code 80001)", () => {
+            const { sourceFile, project } = getInfoFromText("const moment = require('moment'); moment(); ", { filePath: "/file.ts" });
+            const variableDeclaration = sourceFile.getVariableDeclarationOrThrow("moment");
+            const results = project.getLanguageService().getCodeFixesAtPosition(sourceFile.getFilePath(), variableDeclaration.getStart(), variableDeclaration.getEnd(), [80001], {}, {});
+
+            expect(results[0]!.getFixName()).to.equal("convertToEs6Module");
+            expect(results[0]!.getDescription()).to.equal("Convert to ES6 module");
+
+            checkFileTextChanges(results[0]!.getChanges()[0], {
```
Comment (dsherret) [inline src/tests/compiler/tools/languageServiceTests.ts]: Does this only return 1 change? Could you add an assertion for how many results this should return?

**Comment by dsherret** [comment]
@cancerberoSgx The `LanguageService` changes look good, but I'm sorry I keep denying code. Could you please read the contributing instructions for submitting features? That would help us not waste effort. Thanks! https://github.com/dsherret/ts-simple-ast/blob/master/CONTRIBUTING.md#contributing-features

**Comment by cancerberoSgx** [comment]
> @cancerberoSgx The `LanguageService` changes look good, but I'm sorry I keep denying code. Could you please read the contributing instructions for submitting features? That would help us not waste effort. Thanks! https://github.com/dsherret/ts-simple-ast/blob/master/CONTRIBUTING.md#contributing-features

No worries! Will do it tomorrow, that's why I separate it in two different commits. I insist on having this discussions on a PR instead of on an issue since it's much more clear and is not a big deal for me, but in the future I will create an issue first.

File: src/compiler/tools/results/TextChange.ts (hunk 4 above)
Comment (cancerberoSgx) [reply src/compiler/tools/results/TextChange.ts]: Why don't you want to publish the constructor? Currently, SourceFile.applyChanges is very useful. When working with TypeScript plugins and I need to cheat the library to instantiate the tsa TextChange thing sourceFile.applyChanges( new (TextChange as any)(tsTextChanges) )...

File: src/compiler/tools/LanguageService.ts (hunk 3 above)
Comment (cancerberoSgx) [reply src/compiler/tools/LanguageService.ts]: > What do you think about making this a single object parameter with all of these parameters as properties on the object?

I think having a similar API to TypeScript will facilitate things for users, so my opinion is no :)

**Comment by cancerberoSgx** [comment]
I think is ready now, thanks

File: src/compiler/tools/LanguageService.ts [hunk 7]
```diff
@@ -8,7 +8,10 @@ import { Node } from "../ast/common";
 import { SourceFile } from "../ast/module";
 import { FormatCodeSettings, UserPreferences, RenameOptions } from "./inputs";
 import { Program } from "./Program";
-import { DefinitionInfo, EmitOutput, FileTextChanges, ImplementationLocation, RenameLocation, TextChange } from "./results";
+import { DefinitionInfo, EmitOutput, FileTextChanges, ImplementationLocation, RenameLocation, TextChange, DiagnosticWithLocation } from "./results";
+import { TextRange } from "typescript";
```
Comment (dsherret) [inline src/compiler/tools/LanguageService.ts]: This should use the export from the `"../../typescript"` module.

File: src/compiler/tools/LanguageService.ts [hunk 8]
```diff
@@ -230,6 +233,16 @@ export class LanguageService {
         return renameLocations.map(l => new RenameLocation(this.context, l));
     }
 
+    /**
+     * Gets the suggestion diagnostics.
+     * @param fileName - Optional source file to filter by.
+     */
+    getSuggestionDiagnostics(filePathOrSourceFile: SourceFile | string): DiagnosticWithLocation[] {
```
Comment (dsherret) [inline src/compiler/tools/LanguageService.ts]: Sorry, didn't notice this before. This should be in its own PR because it's a different feature. Doing so makes it work well with https://www.conventionalcommits.org (which is how I change the commit message when squashing the PR).

Thanks for adding this though!

File: src/compiler/tools/LanguageService.ts [hunk 9]
```diff
@@ -230,6 +233,16 @@ export class LanguageService {
         return renameLocations.map(l => new RenameLocation(this.context, l));
     }
 
+    /**
+     * Gets the suggestion diagnostics.
+     * @param fileName - Optional source file to filter by.
```
Comment (dsherret) [inline src/compiler/tools/LanguageService.ts]: Doc doesn't match parameter.

File: src/compiler/tools/results/TextChange.ts (hunk 4 above)
Comment (dsherret) [reply src/compiler/tools/results/TextChange.ts]: I often do breaking changes on the constructor.

**Comment by dsherret** [comment]
@cancerberoSgx I'm going to quickly make all the changes and merge this PR. Will have it done in 15-30 mins so don't bother with any changes at the moment.

File: src/compiler/tools/LanguageService.ts (hunk 8 above)
Comment (dsherret) [reply src/compiler/tools/LanguageService.ts]: Actually, this is ok. I'll keep it in the PR because this is all language service changes.

File: src/compiler/tools/LanguageService.ts [hunk 10]
```diff
@@ -337,6 +350,41 @@ export class LanguageService {
...
+     * Gets the edit information to modify the file with given code fix at given text range in a given file.
+     * @param fileName - File path of the source file.
+     * @param start - The start index of the text range to be fixed.
+     * @param end  - The end index of the text range to be fixed.
+     * @param errorCodes  - One or more error codes associated to the code fixex to apply.
+     * @param formatOptions - Format coe settings.
+     * @param preferences . User preferences for refactoring.
+     */
+    getCodeFixesAtPosition(fileName: string, start: number, end: number, errorCodes: ReadonlyArray<number>,
+        formatOptions: FormatCodeSettings, preferences: UserPreferences): ReadonlyArray<CodeFixAction> {
```
Comment (dsherret) [inline src/compiler/tools/LanguageService.ts]: Inputs should be readonly, but outputs like this one should be mutable. That gives people the greatest flexibility since they can provide arrays that are restrictive and then mutate arrays that are given back to them if they wish.

Comment (dsherret) [reply src/compiler/tools/LanguageService.ts]: Oh, I just realized the compiler is doing this. Sometimes it's better not to listen to what they do.

File: src/compiler/tools/results/FileTextChanges.ts [hunk 11]
```diff
@@ -27,4 +27,18 @@ export class FileTextChanges {
...
+     * Should the file be created?
+     */
+    isNewFile() {
+        return this._compilerObject.isNewFile;
+    }
+
+    /**
+     * Creates a new FileTextChange from compiler object.
+     */
+    static create(compilerObject: ts.FileTextChanges) {
```
Comment (dsherret) [inline src/compiler/tools/results/FileTextChanges.ts]: Why does this need to exist?

File: src/compiler/tools/results/RefactorEditInfo.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    @Memoize
+    getEdits() {
+        return this.compilerObject.edits.map(edit => new FileTextChanges(edit));
+    }
+
+    /**
+     * Gets the file path for a rename refactor.
+     */
+    getRenameFilePath() {
```
Comment (dsherret) [inline src/compiler/tools/results/RefactorEditInfo.ts]: Needs a test that accesses this.

File: src/compiler/tools/results/RefactorEditInfo.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        return this.compilerObject.renameFilename;
+    }
+
+    getRenameLocation() {
```
Comment (dsherret) [inline src/compiler/tools/results/RefactorEditInfo.ts]: What does this do? Needs a test that accesses this.

File: src/tests/compiler/statement/statementTests.ts [hunk 14]
```diff
@@ -1,5 +1,14 @@
+import { expect } from "chai";
 import { Statement } from "../../../compiler";
+import { getInfoFromText } from "../testHelpers";
 
 describe(nameof(Statement), () => {
-    // todo: tests
+    describe(nameof<Statement>(n => n.remove), () => {
```
Comment (dsherret) [inline src/tests/compiler/statement/statementTests.ts]: Thanks!

File: src/compiler/tools/results/CodeFixAction.ts [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+     */
+    getFixName() {
+        return this.compilerObject.fixName;
+    }
+
+    /**
+     * If present, one may call 'getCombinedCodeFix' with this fixId.
+     * This may be omitted to indicate that the code fix can't be applied in a group.
+     */
+    getFixId() {
```
Comment (dsherret) [inline src/compiler/tools/results/CodeFixAction.ts]: Add test?

File: src/compiler/tools/results/CodeFixAction.ts [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+        return this.compilerObject.fixId;
+    }
+
+    /**
+     * Gets the description of the code fix when fixing everything.
+     */
+    getFixAllDescription() {
```
Comment (dsherret) [inline src/compiler/tools/results/CodeFixAction.ts]: Add test?

File: src/compiler/tools/results/FileTextChanges.ts [hunk 17]
```diff
@@ -27,4 +25,11 @@ export class FileTextChanges {
     getTextChanges() {
         return this._compilerObject.textChanges.map(c => new TextChange(c));
     }
+
+    /**
+     * Gets if this change is for creating a new file.
+     */
+    isNewFile() {
```
Comment (dsherret) [inline src/compiler/tools/results/FileTextChanges.ts]: Add test?

**Comment by dsherret** [comment]
Hey @cancerberoSgx, this is almost done. I made some updates because it was quicker than reviewing. We just need a few tests for the remaining comments I made before this is merged. I've stopped working on this now so if you have the time to add those tests then please add them. Otherwise, I will get to it sometime next weekend.

**Comment by cancerberoSgx** [comment]
added more test asserts

File: src/compiler/tools/LanguageService.ts [hunk 18]
```diff
@@ -8,7 +8,14 @@ import { Node } from "../ast/common";
 import { SourceFile } from "../ast/module";
 import { FormatCodeSettings, UserPreferences, RenameOptions } from "./inputs";
 import { Program } from "./Program";
-import { DefinitionInfo, EmitOutput, FileTextChanges, ImplementationLocation, RenameLocation, TextChange } from "./results";
+import { DefinitionInfo, EmitOutput, FileTextChanges, ImplementationLocation, RenameLocation, TextChange, DiagnosticWithLocation } from "./results";
+import { RefactorEditInfo } from "./results/RefactorEditInfo";
+import { CodeFixAction } from "./results/CodeFixAction";
```
Comment (dsherret) [inline src/compiler/tools/LanguageService.ts]: These two imports aren't included in the barrel file (`./results/index.ts`). Could you add them there and then fix these imports?

That should fix these CI errors:

```ts
> ts-node --transpileOnly scripts/verification/ensureNoDeclarationFileErrors
dist-declarations/ts-simple-ast.d.ts:8897:214 - error TS2304: Cannot find name 'RefactorEditInfo'.
8897     getEditsForRefactor(filePathOrSourceFile: string | SourceFile, formatOptions: FormatCodeSettings, positionOrRange: number | TextRange, refactorName: string, actionName: string, preferences?: UserPreferences): RefactorEditInfo | undefined;
                                                                                                                                                                                                                          ~~~~~~~~~~~~~~~~
dist-declarations/ts-simple-ast.d.ts:8907:202 - error TS2304: Cannot find name 'CodeFixAction'.
8907     getCodeFixesAtPosition(filePathOrSourceFile: string | SourceFile, start: number, end: number, errorCodes: ReadonlyArray<number>, formatOptions?: FormatCodeSettings, preferences?: UserPreferences): CodeFixAction[];
                                                                                                                                                                                                              ~~~~~~~~~~~~~
There were declaration file issues!
```

File: src/compiler/tools/results/RefactorEditInfo.ts (hunk 13 above)
Comment (dsherret) [inline src/compiler/tools/results/RefactorEditInfo.ts]: Do you know a good JS doc comment for this?

**Comment by dsherret** [comment]
Thanks once again @cancerberoSgx! :)
